### PR TITLE
Support for setting geo-replication clusters on topic level

### PR DESCRIPTION
**Pull request by Technoboy-**
## Motivation
Currently, when we try to enable the geo-replication, we need to set multiple clusters for the namespaces, and all the topics under the namespace will start to replicate data across clusters.

It's useful for adding the replicate clusters in the topic level so that we can achieve the topics under a namespace can have different replication clusters such as topic A(cluster-a <-> cluster-b) topic B(cluster-b <-> cluster-c).

## Documentation
I'm not sure if the doc could generate automatically. If no, I will update the doc later.

### CODE REVIEW

**Comment by Anonymitaet** [comment]
> I'm not sure if the doc could generate automatically. If no, I will update the doc later.

@Technoboy- the doc you added can be generated automatically.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 1]
```diff
@@ -1334,33 +1335,36 @@ public void closeFailed(ManagedLedgerException exception, Object ctx) {
...
-        CompletableFuture<Policies> policiesFuture = brokerService.pulsar().getPulsarResources()
-                .getNamespaceResources()
-                .getPoliciesAsync(TopicName.get(topic).getNamespaceObject())
-                .thenCompose(optPolicies -> {
-                            if (!optPolicies.isPresent()) {
-                                return FutureUtil.failedFuture(
-                                        new ServerMetadataException(
-                                                new MetadataStoreException.NotFoundException()));
-                            }
+        CompletableFuture<Set<String>> replicationClustersFuture = brokerService.pulsar()
```
Comment (hangc0276) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1338]: The checkReplication is call by `onPoliciesUpdate`, which is listen namespace level policy update. For topic level policy update listener, we should deal with the topic policy in `onUpdate()` method.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java [hunk 2]
```diff
@@ -2641,4 +2642,23 @@ public void testDoNotCreateSystemTopicForHeartbeatNamespace() {
         });
     }
 
+    @Test(timeOut = 30000)
```
Comment (hangc0276) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java:2645]: We'd better setup multi individual pulsar clusters, and turn on topic level replication policy for specific topic to test topic replicate policy. 
For example, we setup cluster-A and cluster-B, and configure `persistent://public/default/topic-test` to replicate from cluster-A to cluster-B,  and then write message-1 to topic `persistent://public/default/topic-test` in cluster-A, we can check whether message-1 can be read from topic `persistent://public/default/topic-test` in cluster-B. Furthermore, we and update, remove topic replication policy, check the messages in two clusters.

**Review by hangc0276 — CHANGES_REQUESTED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java [hunk 3]
```diff
@@ -495,23 +495,28 @@ public void removeProducer(Producer producer) {
...
 
-        return brokerService.pulsar().getPulsarResources().getNamespaceResources()
-                .getPoliciesAsync(TopicName.get(topic).getNamespaceObject())
-                .thenCompose(optPolicies -> {
-                    if (!optPolicies.isPresent()) {
-                        return FutureUtil.failedFuture(
-                                new ServerMetadataException(
-                                        new MetadataStoreException.NotFoundException()));
-                    }
+        return brokerService.pulsar()
```
[the remark this answers is not in the document]
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/nonpersistent/NonPersistentTopic.java:498]: Thanks for the kindly review.

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java (hunk 2 above)
Comment (Technoboy-) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java:2645]: Yes, I will create a test to cover this later.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 1 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1338]: From the discussion, my implement and your suggestion both can realize this.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java [hunk 4]
```diff
@@ -2708,6 +2709,60 @@ protected PersistentOfflineTopicStats internalGetBacklog(boolean authoritative)
...
+        );
+    }
+
+    protected CompletableFuture<Void> internalRemoveReplicationClusters() {
+        validateTopicPolicyOperation(topicName, PolicyName.REPLICATION, PolicyOperation.WRITE);
+        validatePoliciesReadOnlyAccess();
+
+        return getTopicPoliciesAsyncWithRetry(topicName).thenCompose(op -> {
+                    TopicPolicies topicPolicies = op.orElseGet(TopicPolicies::new);
+                    topicPolicies.setReplicationClusters(Collections.emptyList());
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java:2752]: emptyList is a static variable. I think we can use list.clear or `new ArrayList()`

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java [hunk 5]
```diff
@@ -226,6 +226,10 @@ public CmdTopics(Supplier<PulsarAdmin> admin) {
 
         jcommander.addCommand("set-replicated-subscription-status", new SetReplicatedSubscriptionStatus());
 
+        jcommander.addCommand("get-replication-clusters", new GetReplicationClusters());
```
Comment (315157973) [inline pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java:229]: Please also add unit test in `PulsarAdminToolTest`

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 6]
```diff
@@ -1334,33 +1335,37 @@ public void closeFailed(ManagedLedgerException exception, Object ctx) {
...
-                                return FutureUtil.failedFuture(
-                                        new ServerMetadataException(
-                                                new MetadataStoreException.NotFoundException()));
-                            }
+        CompletableFuture<List<String>> replicationClustersFuture = brokerService.pulsar()
+                .getTopicPoliciesService()
+                .getTopicPoliciesAsyncWithRetry(name, null, brokerService.pulsar().getExecutor())
+                .thenCompose(topicPolicies -> {
+                    if (!topicPolicies.isPresent()
+                            || CollectionUtils.isEmpty(topicPolicies.get().getReplicationClusters())) {
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:1343]: We need to define the state:
null means no setting
The list is Empty, which means that the value is not set, that is, the copy of this Topic under Namespace is disabled

If empty also means that there is no setting, subsequent Topic-level Policies cannot be disabled

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java (hunk 2 above)
Comment (315157973) [inline pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java:2645]: Please add the following unit tests:
1)The namespace and topic level policies are set at the same time, whether the priority is correct

2)Disabling topic-level policies

File: pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java (hunk 5 above)
Comment (Technoboy-) [reply pulsar-client-tools/src/main/java/org/apache/pulsar/admin/cli/CmdTopics.java:229]: Ok, done

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java (hunk 4 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/impl/PersistentTopicsBase.java]: ok

File: pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java (hunk 2 above)
Comment (Technoboy-) [reply pulsar-broker/src/test/java/org/apache/pulsar/broker/admin/TopicPoliciesTest.java:2645]: done.

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 6 above)
Comment (Technoboy-) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Ok, done.

**Comment by 315157973** [comment]
@hangc0276  Please review again

**Review by codelipenghui — APPROVED** [review]

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 1 above)
Comment (hangc0276) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java]: Current implementation relies on `replicationPolicyCheckDurationSeconds = 600` to check replication policy update. We can merge this PR first, and consider use another PR to update the replication policy by listener.

**Review by hangc0276 — APPROVED** [review]

**Review by 315157973 — APPROVED** [review]

**Pull request merged**